# Preprocessing:

In [18]:
import os
import numpy as np
from PIL import Image
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix
from sklearn.model_selection import KFold
import tensorflow as tf
from tensorflow.keras import layers, models
from tensorflow.keras.optimizers import Adam, SGD
import time


In [2]:
root_dir = "./10000/"

X = []
y = []

In [3]:


for label in range(10):
    folder_path = os.path.join(root_dir, str(label))
    
    for filename in os.listdir(folder_path):
        if filename.endswith(".jpg") or filename.endswith(".jpeg"):
            img_path = os.path.join(folder_path, filename)

          
            img = Image.open(img_path)
            img = img.convert("L")
            img = img.resize((28, 28))
            img_array = np.array(img)
            img_flat = img_array.flatten()

            X.append(img_flat)
            y.append(label)

X = np.array(X)
y = np.array(y)

print("X shape:", X.shape)
print("y shape:", y.shape)

X shape: (10000, 784)
y shape: (10000,)


In [4]:
X = X / 255.0

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

print("Train size:", X_train.shape)
print("Test size:", X_test.shape)

Train size: (8000, 784)
Test size: (2000, 784)


# Deep learning model 1 :

In [10]:
model = models.Sequential([
    layers.Input(shape=(784,)),               
    layers.Dense(128, activation='tanh'),
    layers.Dense(64, activation='tanh'),
    layers.Dense(10, activation='softmax')    
])
optimizer = tf.keras.optimizers.SGD(
        learning_rate=0.01,
)
model.compile(
    optimizer=optimizer,
    loss='sparse_categorical_crossentropy', 
    metrics=['accuracy']
)

history = model.fit(
    X_train, y_train,
    epochs=20,
    batch_size=32,
    validation_split=0.1
)


Epoch 1/20
225/225 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.2231 - loss: 2.1794 - val_accuracy: 0.3663 - val_loss: 2.0153
Epoch 2/20
225/225 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.4318 - loss: 1.8690 - val_accuracy: 0.5238 - val_loss: 1.7146
Epoch 3/20
225/225 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5382 - loss: 1.5891 - val_accuracy: 0.5725 - val_loss: 1.4812
Epoch 4/20
225/225 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5954 - loss: 1.3842 - val_accuracy: 0.5550 - val_loss: 1.3748
Epoch 5/20
225/225 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6265 - loss: 1.2614 - val_accuracy: 0.5537 - val_loss: 1.3650
Epoch 6/20
225/225 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6515 - loss: 1.1652 - val_accuracy: 0.6700 - val_loss: 1.1199
Epoch 7/20
225/225 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6736 - loss: 1.0859 - val_accuracy: 0.6637 - val_loss: 1.1007
Epoch 8/20
225/225 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6737 - loss: 1.0413 - val_accuracy: 0.

In [11]:
y_pred = model.predict(X_test)
y_pred_labels = np.argmax(y_pred, axis=1)

print("Accuracy:", accuracy_score(y_test, y_pred_labels))
print("Precision:", precision_score(y_test, y_pred_labels, average='macro'))
print("Recall:", recall_score(y_test, y_pred_labels, average='macro'))
print("F1-score:", f1_score(y_test, y_pred_labels, average='macro'))

print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred_labels))


63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step
Accuracy: 0.7415
Precision: 0.7685910789032433
Recall: 0.7414999999999999
F1-score: 0.7431973817073517
Confusion Matrix:
[[163   3   0   4   6   6   5   4   6   3]
 [  3 130   3   6   7   8   1  29  13   0]
 [  5   2 115  11   7  18   3  19  12   8]
 [  8   2   4 150   8   6   0   9  11   2]
 [  2   1   1   2 155   4   8  15  10   2]
 [  4   1   0   1   4 167   8   5  10   0]
 [  1   4   0   0   2  20 165   1   7   0]
 [  1   0   0   3   6   2   1 167  15   5]
 [  8   1   0   5  10  14   4  11 144   3]
 [  9   0   0  11  13   0   0  38   2 127]]


# Deep learning model 2 :

In [48]:
model = models.Sequential([
    layers.Input(shape=(784,)),               
    layers.Dense(512, activation='tanh'),
    layers.Dense(512, activation='tanh'),

    layers.Dense(256, activation='tanh'),
    layers.Dense(128, activation='tanh'),
    layers.Dense(64, activation='tanh'),
    layers.Dense(10, activation='softmax')    
])

optimizer = tf.keras.optimizers.SGD(
        learning_rate=0.02,
)
model.compile(
    optimizer=optimizer,
    loss='sparse_categorical_crossentropy', 
    metrics=['accuracy']
)
start_time = time.time()
history = model.fit(
    X_train, y_train,
     epochs=100,
    batch_size=32,
    validation_split=0.1
)
end_time = time.time()
training_time = end_time - start_time

Epoch 1/100
225/225 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.2556 - loss: 2.0737 - val_accuracy: 0.2575 - val_loss: 2.1716
Epoch 2/100
225/225 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.4621 - loss: 1.5974 - val_accuracy: 0.5288 - val_loss: 1.3994
Epoch 3/100
225/225 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.5544 - loss: 1.3599 - val_accuracy: 0.6288 - val_loss: 1.1809
Epoch 4/100
225/225 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.5997 - loss: 1.2372 - val_accuracy: 0.5362 - val_loss: 1.4161
Epoch 5/100
225/225 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.6442 - loss: 1.0953 - val_accuracy: 0.6500 - val_loss: 1.0858
Epoch 6/100
225/225 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.6710 - loss: 1.0315 - val_accuracy: 0.6975 - val_loss: 0.9575
Epoch 7/100
225/225 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.6881 - loss: 0.9604 - val_accuracy: 0.6100 - val_loss: 1.1800
Epoch 8/100
225/225 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.6976 - loss: 0.9313 - val_accu

In [49]:
y_pred = model.predict(X_test)
y_pred_labels = np.argmax(y_pred, axis=1)
print("Training time (seconds):", training_time)
print("Accuracy:", accuracy_score(y_test, y_pred_labels))
print("Precision:", precision_score(y_test, y_pred_labels, average='macro'))
print("Recall:", recall_score(y_test, y_pred_labels, average='macro'))
print("F1-score:", f1_score(y_test, y_pred_labels, average='macro'))

print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred_labels))


63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step
Training time (seconds): 86.27867722511292
Accuracy: 0.8705
Precision: 0.8770399132046824
Recall: 0.8705
F1-score: 0.8698839518686219
Confusion Matrix:
[[189   2   0   0   0   1   4   0   0   4]
 [  0 191   2   0   4   0   1   2   0   0]
 [  0   5 180   2   4   3   0   2   1   3]
 [  1   1   5 171   6   7   0   0   5   4]
 [  1   5   5   0 181   0   3   0   1   4]
 [  3   0   2   0   3 183   4   0   4   1]
 [  3   2   1   0   4   5 183   0   2   0]
 [  2  40   0   0   1   0   0 143   2  12]
 [  7   5   3   3  10   2  10   2 154   4]
 [  6   4   4   7   9   1   0   3   0 166]]


# Deep learning model 3 :

In [14]:
model = models.Sequential([
    layers.Input(shape=(784,)),  
    layers.Dense(256, activation='tanh'),             
    layers.Dense(512, activation='tanh'),
    layers.Dense(512, activation='tanh'),
    layers.Dense(1024, activation='tanh'),
    layers.Dense(512, activation='tanh'),
    layers.Dense(10, activation='softmax')    
])

optimizer = tf.keras.optimizers.SGD(
        learning_rate=0.01,
)
model.compile(
    optimizer=optimizer,
    loss='sparse_categorical_crossentropy', 
    metrics=['accuracy']
)

history = model.fit(
    X_train, y_train,
    epochs=20,
    batch_size=32,
    validation_split=0.1
)

Epoch 1/20
225/225 ━━━━━━━━━━━━━━━━━━━━ 4s 9ms/step - accuracy: 0.3369 - loss: 1.9711 - val_accuracy: 0.4275 - val_loss: 1.6286
Epoch 2/20
225/225 ━━━━━━━━━━━━━━━━━━━━ 2s 7ms/step - accuracy: 0.5203 - loss: 1.4763 - val_accuracy: 0.6075 - val_loss: 1.2097
Epoch 3/20
225/225 ━━━━━━━━━━━━━━━━━━━━ 2s 7ms/step - accuracy: 0.5624 - loss: 1.3185 - val_accuracy: 0.5875 - val_loss: 1.2396
Epoch 4/20
225/225 ━━━━━━━━━━━━━━━━━━━━ 1s 6ms/step - accuracy: 0.6199 - loss: 1.1854 - val_accuracy: 0.6637 - val_loss: 1.0401
Epoch 5/20
225/225 ━━━━━━━━━━━━━━━━━━━━ 1s 6ms/step - accuracy: 0.6481 - loss: 1.1145 - val_accuracy: 0.6212 - val_loss: 1.1805
Epoch 6/20
225/225 ━━━━━━━━━━━━━━━━━━━━ 2s 7ms/step - accuracy: 0.6589 - loss: 1.0533 - val_accuracy: 0.6612 - val_loss: 1.0727
Epoch 7/20
225/225 ━━━━━━━━━━━━━━━━━━━━ 1s 6ms/step - accuracy: 0.6775 - loss: 1.0137 - val_accuracy: 0.5950 - val_loss: 1.3156
Epoch 8/20
225/225 ━━━━━━━━━━━━━━━━━━━━ 1s 6ms/step - accuracy: 0.6851 - loss: 0.9847 - val_accuracy: 0.

In [15]:
y_pred = model.predict(X_test)
y_pred_labels = np.argmax(y_pred, axis=1)

print("Accuracy:", accuracy_score(y_test, y_pred_labels))
print("Precision:", precision_score(y_test, y_pred_labels, average='macro'))
print("Recall:", recall_score(y_test, y_pred_labels, average='macro'))
print("F1-score:", f1_score(y_test, y_pred_labels, average='macro'))

print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred_labels))


63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step
Accuracy: 0.7665
Precision: 0.7744106453380601
Recall: 0.7665
F1-score: 0.7652853774147319
Confusion Matrix:
[[136  12   5   6   5  13  12   1   4   6]
 [  1 157   9   5   0   5   7  11   5   0]
 [  1   3 170   9   3   1   2   5   4   2]
 [  3   3  11 163   2   4   1   2  10   1]
 [  1   4   8   7 136   4  13   9  12   6]
 [  1   3   4   3   1 161  17   3   7   0]
 [  0   3   0   0   0   8 186   0   3   0]
 [  0   7  15   6   3   2   4 152   5   6]
 [  5   5   7  10   7  13  11   5 133   4]
 [  3   2  12  12   9   1   0  21   1 139]]


# Model 2 tunning:

## with learning_rate=0.01 and 20 epoch and 32 batch size: for all 3 models:
training : accuracy: 0.7806 
testing : Accuracy: 0.785

## with learning_rate=0.02 and 20 epoch and 32 batch size:
training : accuracy: 0.8096 
testing : Accuracy: 0.791
شويه overfit

## with learning_rate=0.02 and 70 epoch and 32 batch size:
training : accuracy: 0.9604 
testing : Accuracy: 0.86
overfit كبير

## with learning_rate=0.02 and 80 epoch and 32 batch size and 4 dropout layers:
training : accuracy: 0.9833
testing : Accuracy: 0.873
overfit كبير

## with learning_rate=0.02 and 100 epoch and 32 batch size :
training : accuracy: 0.9929 
testing : Accuracy: 0.8975
overfit كبير

# k_fold for Deep Learning model :

In [50]:
def create_model():
    model = models.Sequential([
        layers.Input(shape=(784,)),               
        layers.Dense(512, activation='tanh'),
        layers.Dense(512, activation='tanh'),

        layers.Dense(256, activation='tanh'),
        layers.Dense(128, activation='tanh'),
        layers.Dense(64, activation='tanh'),
        layers.Dense(10, activation='softmax')    
    ])

    optimizer = tf.keras.optimizers.SGD(
        learning_rate=0.02,
    )
    model.compile(
        optimizer=optimizer,
        loss='sparse_categorical_crossentropy', 
        metrics=['accuracy']
    )
    return model


In [51]:
kf_nn = KFold(n_splits=5, shuffle=True, random_state=42)

cv_scores_nn = []

for train_idx, test_idx in kf_nn.split(X):
    X_train_fold, X_test_fold = X[train_idx], X[test_idx]
    y_train_fold, y_test_fold = y[train_idx], y[test_idx]

    model = create_model()

    model.fit(
        X_train_fold, y_train_fold,
        epochs=100,
        batch_size=32,
        validation_split=0.1
    )

    loss, fold_acc = model.evaluate(
        X_test_fold, y_test_fold
    )

    print(f"Fold accuracy: {fold_acc:.4f}")
    cv_scores_nn.append(fold_acc)

print("Mean :", np.mean(cv_scores_nn))
print("Standard deviation :", np.std(cv_scores_nn))


Epoch 1/100
225/225 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.2804 - loss: 1.9704 - val_accuracy: 0.0000e+00 - val_loss: 4.8232
Epoch 2/100
225/225 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.4829 - loss: 1.5157 - val_accuracy: 0.0000e+00 - val_loss: 4.4990
Epoch 3/100
225/225 ━━━━━━━━━━━━━━━━━━━━ 1s 4ms/step - accuracy: 0.5822 - loss: 1.2669 - val_accuracy: 0.0000e+00 - val_loss: 4.7988
Epoch 4/100
225/225 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.6326 - loss: 1.1320 - val_accuracy: 0.0000e+00 - val_loss: 5.1050
Epoch 5/100
225/225 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.6657 - loss: 1.0425 - val_accuracy: 0.0000e+00 - val_loss: 5.2704
Epoch 6/100
225/225 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.6951 - loss: 0.9474 - val_accuracy: 0.0000e+00 - val_loss: 4.7720
Epoch 7/100
225/225 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.7043 - loss: 0.9129 - val_accuracy: 0.0000e+00 - val_loss: 5.2622
Epoch 8/100
225/225 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.73

In [52]:
print(cv_scores_nn)

[0.784500002861023, 0.8125, 0.8040000200271606, 0.8054999709129333, 0.8174999952316284]
